src/relations/db.py

```diff
@@ -88,113 +130,243 @@ def __init__(self, charm: CharmBase, admin: bool = False):
         self.charm = charm
         self.admin = admin
 
-    def _on_relation_changed(self, change_event: RelationChangedEvent):
-        """Handle db-relation-changed event.
+    def _on_relation_joined(self, join_event: RelationJoinedEvent):
+        """Handle db-relation-joined event.
 
-        Takes information from the db relation databag and copies it into the pgbouncer.ini
-        config.
+        If the backend relation is fully initialised and available, we generate the proposed
+        database and create a user on the postgres charm, and add preliminary data to the databag.
         """
-        if not self.charm.unit.is_leader():
+        if not self.charm.backend.postgres:
+            # We can't relate an app to the backend database without a backend postgres relation
+            wait_str = "waiting for backend-database relation to connect"
+            logger.warning(wait_str)
+            self.charm.unit.status = WaitingStatus(wait_str)
+            join_event.defer()
+            return
+
+        try:
+            cfg = self.charm.read_pgb_config()
+        except FileNotFoundError:
+            wait_str = "waiting for pgbouncer to start"
+            logger.warning(wait_str)
+            self.charm.unit.status = WaitingStatus(wait_str)
+            join_event.defer()
             return
 
-        logger.info(f"Setting up {change_event.relation.name} relation - updating config")
+        logger.info(f"Setting up {self.relation_name} relation")
         logger.warning(
             f"DEPRECATION WARNING - {self.relation_name} is a legacy relation, and will be deprecated in a future release. "
         )
 
+        remote_databag = join_event.relation.data[join_event.app]
+
+        database = remote_databag.get("database")
+        if not database:
+            # If the database field isn't populated, the remote app may be using the unit databag.
+            remote_databag = join_event.relation.data[join_event.unit]
+            database = remote_databag.get("database")
+
+            if not database:
+                # If there's nothing in either databag, break.
+                logger.warning("No database name provided in app databag")
+                join_event.defer()
+                return
```

Comment:
Minor simiplication to help with indention complexity.

Followup questions: 

1. Why might it be provided in either app or unit data? Is that something we can control, because it would be nice to have consistency.
2. Should we go into WaitingStatus or is that overkill?

```suggestion
        remote_app_databag = join_event.relation.data[join_event.app]
        remote_unit_databag = join_event.relation.data[join_event.unit]

        database = remote_app_databag.get("database")
        # If the database field isn't populated, the remote app may be using the unit databag.
        database = remote_unit_databag.get("database") if not databag
        if not database:
             # If there's nothing in either databag, break.
             logger.warning("No database name provided in app or unit databag")
             join_event.defer()
             return
```

Reply:
1. One of the databags (I believe it was app) was added after the unit databag, so different charms handle it differently depending on when they were written, and there's no guarantee which one is used because the database name is provided by the client charm. After this point, we spam our own databags with all the info we need, because client charms can't be trusted. 
2. Waitingstatus should be fine, I've added it in. 

Good suggestion though, I've tightened it up further using some walrus operators.